src/common/src/util/resource_util.rs

```diff
@@ -0,0 +1,625 @@
+    const DEFAULT_DOCKER_ENV_PATH: &str = "/.dockerenv";
+    const DEFAULT_LINUX_IDENTIFIER: &str = "linux";
+    const DEFAULT_IN_CONTAINER_ENV_VARIABLE: &str = "IN_CONTAINER";
+    const DEFAULT_KUBERNETES_SECRETS_PATH: &str = "/var/run/secrets/kubernetes.io";
+
+    pub fn is_linux_machine() -> bool {
+        env::consts::OS.eq(DEFAULT_LINUX_IDENTIFIER)
+    }
+
+    // checks if is running in a docker container by checking for docker env file, or if it is
+    // running in a kubernetes pod.
+    pub fn is_running_in_container() -> bool {
+        env_var_check_if_running_in_container()
+            || docker_env_exists()
+            || is_running_in_kubernetes_pod()
+    }
+
+    // checks for existence of docker env file
+    pub fn docker_env_exists() -> bool {
+        Path::new(DEFAULT_DOCKER_ENV_PATH).exists()
+    }
+
+    // checks for environment
+    pub fn env_var_check_if_running_in_container() -> bool {
+        env::var(DEFAULT_IN_CONTAINER_ENV_VARIABLE).is_ok()
+    }
+
+    // checks if it is running in a kubernetes pod
+    pub fn is_running_in_kubernetes_pod() -> bool {
+        Path::new(DEFAULT_KUBERNETES_SECRETS_PATH).exists()
+    }
+}
+
+pub mod memory {
+    use sysinfo::{System, SystemExt};
+
+    // Default paths for memory limtiations and usage for cgroup_v1 and cgroup_v2.
+    const V1_MEMORY_LIMIT_HIERARCHY: &str = "/memory/memory.limit_in_bytes";
+    const V1_MEMORY_CURRENT_HIERARCHY: &str = "/memory/memory.usage_in_bytes";
+    const V2_MEMORY_LIMIT_HIERARCHY: &str = "/memory.max";
+    const V2_MEMORY_CURRENT_HIERARCHY: &str = "/memory.current";
+
+    // Returns the system memory.
+    fn get_system_memory() -> usize {
+        let mut sys = System::new();
+        sys.refresh_memory();
+        sys.total_memory() as usize
+    }
+
+    // Returns the used memory of the system.
+    fn get_system_memory_used() -> usize {
+        let mut sys = System::new();
+        sys.refresh_memory();
+        sys.used_memory() as usize
+    }
+
+    /// Returns the total memory used by the system in bytes.
+    ///
+    /// If running in container, this function will read the cgroup interface files for the
+    /// memory used, if interface files are not found, will return the memory used in the system
+    /// as default. The cgroup mount point is assumed to be at /sys/fs/cgroup by default.
+    ///
+    /// [`with_capacity`]: #method.with_capacity
+    ///
+    /// # Examples
+    ///
+    /// Basic usage:
+    /// ``` ignore
+    /// let mem_used = memory::total_memory_used_bytes();
+    /// ```
+    pub fn total_memory_used_bytes() -> usize {
+        if !super::runtime::is_linux_machine() || !super::runtime::is_running_in_container() {
+            return get_system_memory_used();
+        };
+        std::cmp::min(
+            get_memory_used_in_container(super::util::get_cgroup_version())
+                .unwrap_or_else(get_system_memory_used),
+            get_system_memory_used(),
+        )
```

Comment:
Looks like it will run the `get_system_memory_used` twice when `get_memory_used_in_container` returns a `None. `